feat: support stream generate for transformers backend

dblate opened the pull request:
## Description
Support stream generate in /v1/chat/completions API for transformers backend.

First version commit, it will conflict with `InferenceStatus` in `transformers_backend.py`, maybe we can discuss how to merge the two streamer, and any other suggestions will be apprecaited.

## Motivation
As described in #176 , it's a common feature in chat API to support stream generate, this pr aims to add this feature.
close #176 .

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

## Checklist
- [x] I have read the [CONTRIBUTION](https://github.com/ServerlessLLM/ServerlessLLM/blob/main/CONTRIBUTING.md) guide.
- [ ] I have updated the tests (if applicable).
- [ ] I have updated the documentation (if applicable).

MartinRepo commented on sllm/serve/backends/transformers_backend.py:
```diff
@@ -224,6 +228,51 @@ def get_detailed_instruct(task_description: str, query: str) -> str:
         return response
 
+    def generate_text(self, inputs, streamer, temperature, max_new_tokens):
+        self.model.generate(
```
`with torch.no_grad()` to save GPU memory

dblate replied:
fixed in the latest commit.

MartinRepo reviewed (changes requested):
I've reviewed the changes, and overall, everything looks well-structured and in good shape from my perspective.

Could you also update the PR description to include details on how to use streaming output? That would help clarify its implementation for future reference.

Additionally, could you provide more details on the conflict with `InferenceStatus` in `transformers_backend.py`? From what I see, `InferenceStatus` extends `BaseStreamer` and includes state management, while `TextIteratorStreamer` doesn't currently support that. If merging the two streamers is the goal, let's discuss how we can best integrate these two streamers.

Looking forward to your thoughts!

dblate commented on sllm/serve/openai_api_protocol.py:
```diff
@@ -0,0 +1,244 @@
+    allow_fine_tuning: bool = False
+    organization: str = "*"
+    group: Optional[str] = None
+    is_blocking: str = False
```
fixed in the latest commit.

dblate commented on sllm/serve/openai_api_protocol.py:
```diff
@@ -0,0 +1,244 @@
+    top_p: Optional[float] = 1.0
+    top_k: Optional[int] = -1
+    n: Optional[int] = 1
+    max_tokens: Optional[int] = None
```
fixed in the latest commit.

dblate commented on sllm/serve/backends/transformers_backend.py:
```diff
@@ -224,6 +228,51 @@ def get_detailed_instruct(task_description: str, query: str) -> str:
+                    yield token
+                break
+            except Empty:
+                await asyncio.sleep(0.01)
```
Yes, I have thought  about this too, for example why not use 0.1, here are some of my concerns:
- Why sleep: TextIteratorStreamer will raise Empty exception when the model has not generated next token, in this situation we need to wait so I use sleep, but time.sleep will block current thread, so use asyncio.sleep instead
- Why 0.01 instead of 0.1: 0.1s means 100ms, it is a significant latency in TPOT(time between consecutive tokens), actually current thread's main task is to generate tokens, asyncio.sleep just give it a chance to handle other things

So overall 0.01 try to minimize token generate latency and give this thread a chance to handle other tasks.

dblate commented:
Thanks for your suggestions,  I have updated the PR description to describe the stream mode usage and implementation details, and updated the code followed the review opinions. And there is one code comment left in discuss which I have explained my thoughts, looking forward to your opinion.

As for streamer, I think merge `InferenceStatus` and `TextIteratorStreamer` maybe a solution to compatible with these two scenarios. I will take a detailed look at how to implement it.
